src/array/cpu/spmm.h

```diff
@@ -57,9 +60,90 @@ void SpMMSumCsr(
         const DType *rhs_off =
             Op::use_rhs ? W + eid * rhs_dim + rhs_add : nullptr;
         out_off[k] += Op::Call(lhs_off, rhs_off);
+          }
+       }
+    }
+  }
+};
+
+/*!
+ * \brief CPU kernelCore of SpMM<float> specialization CopyLhs.
+ * \param bcast Broadcast information.
+ * \param csr The Csr matrix.
+ * \param ufeat The feature on source nodes.
+ * \param efeat The feature on edges.
+ * \param out The result feature on destination nodes.
+ * \note it uses node parallel strategy, different threads are responsible
+ *       for the computation of different nodes.
+ */
+#if !defined(_WIN32)
+template <class IdType>
+struct SpMMSumCsrCore<IdType, float, dgl::aten::cpu::op::CopyLhs<float>> {
+        static inline void call(const BcastOff &bcast,
+                                const CSRMatrix &csr,
+                                NDArray ufeat, NDArray efeat,
+                                NDArray out) {
+          typedef float DType;
+          const IdType *indptr = csr.indptr.Ptr<IdType>();
+          const IdType *indices = csr.indices.Ptr<IdType>();
+          const DType *X = ufeat.Ptr<DType>();
+          int64_t dim = bcast.out_len,
+                  lhs_dim = bcast.lhs_len;
+          DType *O = out.Ptr<DType>();
+          typedef intel::elem_wise_add_update<DType> ElemWiseUpd;
+
+          /* Prepare an assembler code for this paricular dimension */
+          thread_local intel::Uptr<ElemWiseUpd> th_cpu_spec((intel::IntelKernel<DType>::enabled())
+           ? new ElemWiseUpd(dim) : nullptr);
+          /* Check if the code has been created */
+          ElemWiseUpd *cpu_spec = (th_cpu_spec && th_cpu_spec->is_applicable())
+           ? th_cpu_spec.get() : nullptr;
+          if (cpu_spec && cpu_spec->require_new_instance<bool>(dim)) {
+            th_cpu_spec.reset(new ElemWiseUpd(dim));
+            cpu_spec = (th_cpu_spec && th_cpu_spec->is_applicable()) ? th_cpu_spec.get() : nullptr;
+          }
```

Comment:
Please change to use our own ThreadLocalStore. You could refer to the example here: https://github.com/dmlc/dgl/blob/master/src/runtime/cuda/cuda_common.h#L75

1. In the header file `intel_cpu_support.h`, define a structure such as `IntelCPUThreadEntry` with a static function returning a pointer of `IntelCPUThreadEntry`.
    ```c++
    struct IntelCPUThreadEntry {
      // any members you'd like to include
      // get the threadlocal instance
      static IntelCPUThreadEntry* ThreadLocal();
    };
    ```
1. In the a separate .cc file, include `<dmlc/thread_local.h>` and define
    ```c++
    typedef dmlc::ThreadLocalStore<IntelCPUThreadEntry> IntelCPUThreadStore;
    IntelCPUThreadEntry* IntelCPUThreadEntry::ThreadLocal() {
      return IntelCPUThreadStore::Get();
    }
    ```
    See example [here](https://github.com/dmlc/dgl/blob/master/src/runtime/cuda/cuda_device_api.cc#L200).
1. To use the thread local object, just include `intel_cpu_support.h` and call `ThreadLocal()`. See example [here](https://github.com/dmlc/dgl/blob/2c04ecb55a121587f6fc6beb6ce1bb593f7bccc1/src/array/cuda/csr2coo.cu#L25).

The logic for compiling new kernels should be as follows:
1. Check whether the CPU supports AVX512 (i.e., `is_applicable`). The check should not be related to the feature dim.
1. Check whether `dim` is larger than `min_dim`.
1. Check whether kernel for `dim` has been generated. If not, compile the kernel.

I can see 1 and 2 being implemented but not 3. Does `Xbyak::CodeGenerator` automatically cache the kernel of the same `dim`? If not, we should maintain a dictionary in the `IntelCPUThreadEntry` ourselves.

Reply:
N/A  the concept  has been changed